Second try to vectorize sample_posterior_predictive.

rpgoldman opened the pull request:
Not ready for prime time; sharing for interested parties.

Replacement for #3548

rpgoldman committed:
Add point_list_to_multitrace()

This is intended to make work on sample posterior predictive easier by making the arguments it takes more predictable.

rpgoldman committed:
Redirect sample_posterior_predictive to other file.

Redirect top level calls, and redirect recursive calls to draw_values that occur inside sample_posterior_predictive.

rpgoldman committed:
Type declarations and docstring typo fix.

rpgoldman committed:
First draft of sample_posterior_predictive rewrite.

This is work in progress only -- there are issues with recursive calls to `draw_values()` that occur inside calls to `sample_posterior_predictive`.

rpgoldman committed:
Removed or edited for unsupported features.

The new version of `sample_posterior_predictive()` does not support the `size` argument, so tests using it are commented out.
Also, we don't support asking for a number of samples that is not a round multiple of the number of chains, so in some cases I bumped the number of samples to fit the input trace object.

rpgoldman committed:
Multiple fixes.

In general, the recursive calls to `draw_values` were not handled properly because I had bugs in the way I handled `ContextManager` objects. I believe those bugs are all fixed now.

rpgoldman committed:
Fix application of theano functions.

Needed to change how theano functions were computed in vectorized variant of spp.

rpgoldman committed:
Fix bug in point_list_to_multitrace.

rpgoldman committed:
Fixed incomplete branching error.

lucianopaz commented on pymc3/backends/ndarray.py:
```diff
@@ -340,3 +340,19 @@ def _slice_as_ndarray(strace, idx):
 
     return sliced
+
+def point_list_to_multitrace(point_list: List[Dict[str, np.ndarray]], model: Optional[Model]) -> MultiTrace:
```
```suggestion
def point_list_to_multitrace(point_list: List[Dict[str, np.ndarray]], model: Optional[Model]=None) -> MultiTrace:
```

lucianopaz commented on pymc3/tests/test_sampling.py:
```diff
@@ -281,19 +282,21 @@ def test_normal_vector(self, caplog):
             # test list input
             ppc0 = pm.sample_posterior_predictive([model.test_point], samples=10)
-            ppc = pm.sample_posterior_predictive(trace, samples=10, var_names=[])
+            ppc = pm.sample_posterior_predictive(trace, samples=12, var_names=[])
```
Why was this change necessary?

rpgoldman replied:
This is necessary for now because I rely on the trace argument to have the right length for sampling.
Since the trace is always a multi-trace, I demand that the `samples` argument be a multiple of the number of chains in the `MultiTrace`. If I knew how to take a `MultiTrace` and smash all of the component chains together into a single chain, this would not be necessary.
If you know how to do that, that would be great.

lucianopaz replied:
The `get_values` method combines all chains together for a single variable. That's why `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` would give you a single dictionary, the keys would be the variable names and the values would be all the samples from every chain in the trace. Wouldn't this dictionary be tracelike?

rpgoldman replied:
> The `get_values` method combines all chains together for a single variable. That's why `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` would give you a single dictionary, the keys would be the variable names and the values would be all the samples from every chain in the trace. Wouldn't this dictionary be tracelike?

Oh, excellent! I will look at that. The only remaining issue would be dealing with the absence of `trace.varnames`, and the fact that it doesn't support a `points` but that would be easy to deal with building on UserDict.  I'll see about using `get_values` and removing the limitation on allowable `samples` values.  That will be an easier canonical representation than the `MultiTrace`.

BTW, I am down to 2 tests failing, and of those 2, one would be fixed by this change and one is a multivariate issue which AFAICT simply isn't fixable by me.

Speaking of which, is there some test which would detect any and all multivariate distributions so that I could make a new `fast_posterior_predictive` yield a *useful* error message, instead of crashing in the middle of `random`?

lucianopaz commented on pymc3/tests/test_sampling.py:
```diff
@@ -281,19 +282,21 @@ def test_normal_vector(self, caplog):
-            assert "a" in ppc
-            assert ppc["a"].shape == (10, 4, 2)
+            # I have removed the size argument.  It's confusing, and
+            # it's not clear what it's for. [2019/08/19:rpg]
```
The `size` argument lets you draw more than one sample per point in the trace. In my opinion, this is necessary when you want to draw posterior predictives samples using a single point estimate, like the MAP or MLE.

rpgoldman replied:
Maybe that's true, but that requires handling *both* `size` *and* `samples`, and the necessary infrastructure in terms of `draw_value`, and `random` does not exist for this, as far as I can tell.
Also, the use of `size` somehow changes the shape of the returned traces, which is a further complication, an added circle in shape hell.
I'd be happy to restore this if I could see a way to do it. But I can't right now.

lucianopaz reviewed (commented):
Thanks @rpgoldman. I left a couple of minor comments written down.

I'm still very hesitant about this PR because of the amount of code duplication and added codebase complexity that it brings with `posterior_predictive.py`. Ideally, we would want to just make a minor change to `sample_posterior_predictive`, leaving most of the `draw_values`, `_draw_value`, etc parts as they are. The ideal change would be to use the current `sample_posterior_predictive` but instead of passing each `point` in the trace in a `for` loop, we would pass a `point={trace.get_values(rv_name) for rv_name in trace.varnames}`. This would pass the entire trace all at once and `draw_values` should have to be able to deal with everything else. What prevents us from doing this now are the multivariate distributions, that are not properly vectorized. I propose that we do something different like:

1. We create a small function in `sampling.py` (lets call it `fast_sample_posterior_predictive`)
2. This function would be almost a complete copy of the original `sample_posterior_predictive`. The only change would be that we wouldn't loop over the points in the trace one at a time, but create a dictionary `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` and pass that as the point.
3. Models with multivariate distributions would fail at this until we finish vectorizing the multivariates properly.

rpgoldman committed:
Check zero-length return from spp with no variables.

rpgoldman committed:
Add theano.gof.graph.Constant handling.

Incorporate fix from @lucianopaz (98635eb5dd555befa8ba0ee2a965cce5d47cc3e8) and made corresponding changes to posterior_predictive.

rpgoldman committed:
More ignores.

rpgoldman commented:
> Thanks @rpgoldman. I left a couple of minor comments written down.
> 
> I'm still very hesitant about this PR because of the amount of code duplication and added codebase complexity that it brings with `posterior_predictive.py`. 

I agree about that, but I want to do this in a "hill-climbing" way: that is, I don't want to break any *other* parts of PyMC3 while I am getting this fixed. So I would rather have duplicate code, at least for now, than break prior predictive sampling, or other such code, on my branch.

So I am generally in agreement about your remarks below, but I'm not sure how to get from here to there, especially given the problems with multivariate distributions.

> Ideally, we would want to just make a minor change to `sample_posterior_predictive`, leaving most of the `draw_values`, `_draw_value`, etc parts as they are. The ideal change would be to use the current `sample_posterior_predictive` but instead of passing each `point` in the trace in a `for` loop, we would pass a `point={trace.get_values(rv_name) for rv_name in trace.varnames}`. This would pass the entire trace all at once and `draw_values` should have to be able to deal with everything else. What prevents us from doing this now are the multivariate distributions, that are not properly vectorized.

I think it's more than just that -- there are also at least two further complications, IIUC: 

1. `sample_posterior_predictive` uses *both* `size` *and* `samples`, and as far as I can tell the other APIs don't support both.  And what I mean by that is that generate samples and random methods don't support them, which would impose substantial pain.  Why not remove it, as I have done, and just sample repeatedly as needed?  I.e., control entirely through `samples` instead? Especially since introducing `size` changes the shape of the output?

2. I don't understand to what extent the theano code relies on having the right input shapes. That's why you will see that I have been slicing the input trace so that its length matches the number of samples (and this accounts for my demanding that the `samples` argument be a multiple of the number of chains in the trace.

> I propose that we do something different like:
> 
>     1. We create a small function in `sampling.py` (lets call it `fast_sample_posterior_predictive`)
> 
>     2. This function would be almost a complete copy of the original `sample_posterior_predictive`. The only change would be that we wouldn't loop over the points in the trace one at a time, but create a dictionary `{rv_name: trace.get_values(rv_name) for rv_name in trace.varnames}` and pass that as the point.

I think I needed to do some things in the guts of `draw_value` to make it work with vectorization, including changing the code that makes inferences about shape, but those are probably limited.  If necessary, maybe we could have a `trace` argument in addition to a `point` argument and branch on which is passed.

I like the idea of having a dictionary instead of a `MultiTrace`, and I could more towards that, once all the tests are passing.  Especially since it would remove the multiple chain issues and the issues from building the `MultiTrace` by hand.
> 
>     3. Models with multivariate distributions would fail at this until we finish vectorizing the multivariates properly.

This would be a new limitation, right?  I.e., `sample_posterior_predictive` works on multivariates now, without vectorization, but won't when it's vectorized instead of point at a time?

rpgoldman commented:
@lucianopaz Maybe as a step towards reducing the duplication of code, we could try to identify the differences between `posterior_predictive_draw_values` and `draw_values` with an eye to bringing them back together?
I guess this would be easier if we could also vectorize `sample_prior_predictive`. But if we don't do that -- and if we cannot vectorize sampling from multivariate distributions, I don't see how we can avoid code duplication.
One thing that's difficult is that I haven't found a good substitute for my use of `contextvars` to override `draw_values` with `posterior_predictive_draw_values`.  I hadn't realized that this was 3.7 specific.  I tried replacing the context variable with a global variable, but it didn't work -- I'm not sure if that's due to a threading issue or what.

rpgoldman committed:
Replace MultiTrace with new _TraceDict.

Created a new _TraceDict class that represents a trace as a dictionary of variable names to numpy arrays, and use that as the internal representation, per @lucianopaz suggestion.
Made this a class so that it can have varnames and a length, like the MultiTrace.
Added a couple of translation tests.

rpgoldman committed:
Rename new version fast_sample_posterior_predictive.

Per @lucianopaz suggestion, I have renamed my new version as fast_sample_posterior_predictive(), and restored the old version.
This is not just nice for testing, but also is critical because we cannot currently vectorize the sampling for multivariate distributions.

rpgoldman committed:
Fix _TraceDict indexing bug.

rpgoldman committed:
Fix test typo.

rpgoldman committed:
Unpleasant patch to deal with wrap_random_with_dist_shape

rpgoldman committed:
Refined test_sample while debugging.

More carefully check the results of posterior predictive sampling with pm.Data.

rpgoldman committed:
Ugly fix for shape problem.

When sampling fast_posterior_predictive gets some values with a degenerate extra dimension (e.g., 1 row of 200 values, instead of just 200 values: (1, 200) instead of (200,) ).  Trim these off.
It would be better if I could figure out where these are coming from and avoid that.

rpgoldman committed:
Add typing_extensions to reqs.

rpgoldman committed:
Fix test bugs.

rpgoldman committed:
Add contextvars requirement.

The contextvars library is used to make the vectorized version of sample_posterior_predictive compatible with the legacy version.
contextvars was added in python 3.7, but there is a compatibility library for 3.6
This is an attempt to ensure it will be loaded in python 3.6

rpgoldman committed:
Test correct behavior of context stack.

rpgoldman committed:
Fix forward reference error.

Local class declared after reference.

rpgoldman committed:
Merge from master.

rpgoldman committed:
Fixed incorrect reshaping.

I believe that this reshaping -- which flattened the samples -- was a hold-over from the pointwise variant.  It was causing test failures.

fonnesbeck commented on requirements-dev.txt:
```diff
@@ -16,3 +16,4 @@ recommonmark>=0.4.0
 seaborn>=0.8.1
 sphinx-autobuild==0.7.1
 sphinx>=1.5.5
+typing-extensions>=3.7.4
```
This is already in the requirements.txt changes below

fonnesbeck commented on pymc3/smc/smc_utils.py:
```diff
@@ -0,0 +1,301 @@
+"""
```
Is this a rebase problem?

rpgoldman replied:
Yes, I believe so -- I think that when git ran `diff` it decided, mistakenly, that I had added this whole file.

Thanks for catching that, deleting.

fonnesbeck commented on .pycheckers:
```diff
@@ -0,0 +1,2 @@
+[DEFAULT]
+pylint_rcfile = .pylintrc
```
newline

rpgoldman replied:
This file simply shouldn't be there -- it's a configuration file used by emacs, and should be ignored. I'll fix that, too.

rpgoldman committed:
Amplify the documentation string.

rpgoldman commented:
I believe that this is ready for merging now.  Ordinarily, I would recommend squash merging, but in this case the history is probably of interest, so I recommend a normal merge.  I _did_ edit the history to make it clearer.

rpgoldman committed:
Add entry to changelog.

twiecki commented on pymc3/distributions/.dir-locals.el:
```diff
@@ -0,0 +1,5 @@
+;;; Directory Local Variables
```
Accidental commit.

rpgoldman replied:
Thanks for catching that.  Fixed in rebase, so it's removed from the history (wait for force push).

twiecki commented on pymc3/distributions/__init__.py:
```diff
@@ -168,5 +170,6 @@
            'Rice',
-           'Simulator'
+           'Simulator',
+           'fast_sample_posterior_predictive'
```
```suggestion
           'fast_sample_posterior_predictive',
```

twiecki commented on pymc3/distributions/posterior_predictive.py:
```diff
@@ -0,0 +1,662 @@
+# TraceIsh = Union[MultiTrace, Dict[str, np.ndarray], List['TraceIsh']]
+Point = Dict[str, np.ndarray]
+
+# def _canonicalize_traceish(trace: TraceIsh) -> Union[MultiTrace, Dict[str, np.ndarray]]:
```
Should this be here?

rpgoldman replied:
Removed, thanks.

twiecki commented on pymc3/distributions/posterior_predictive.py:
```diff
@@ -0,0 +1,662 @@
+                    else:
+                        self.named_nodes_children[k].update(nnc[k])
+
+    def draw_value(self, param, trace: Optional[_TraceDict]=None, givens = None):
```
```suggestion
    def draw_value(self, param, trace: Optional[_TraceDict]=None, givens=None):
```

rpgoldman replied:
Thanks for catching that: not sure why pylint didn't complain.

rpgoldman committed:
Remove dead code.

rpgoldman committed:
Style fix.

Co-Authored-By: Thomas Wiecki <thomas.wiecki@gmail.com>

rpgoldman commented:
@twiecki I have addressed all your comments, for which many thanks.  New version pushed and tests are running.  Also rebased on master, for a clean merge.

rpgoldman commented:
This test shows up as failing, but it's a timeout error -- there are no actually failed tests.

lucianopaz commented on pymc3/pytestdebug.log:
```diff
@@ -0,0 +1,1265 @@
+versions pytest-5.0.1, py-1.8.0, python-3.7.4.final.0
```
This log file shouldn't be here. Maybe you could add a rule to .gitignore

rpgoldman replied:
I will remove it from the history.  There's already a rule for it in .gitignore.

rpgoldman replied:
Removed it from the history.  There's a rule for it in .gitignore.   I must have accidentally committed it in the blob immediately before the blob where I committed the rule in .gitignore.

lucianopaz commented on pymc3/tests/test_data_container.py:
```diff
@@ -40,19 +40,30 @@ def test_sample(self):
 
         assert prior_trace0['b'].shape == (1000,)
         assert prior_trace0['obs'].shape == (1000, 100)
+        assert prior_trace1['obs'].shape == (1000, 200)
```
Why does this have a different shape than the one above?

rpgoldman replied:
This happens because the shape of the predictor variable (`x_shared`) has changed.  It was originally a vector of size 100 (line 36), but on line 45, we update it to be of size 200 (see line 33 for the new value).  This changes the size of the prior predictive trace.

This was always true, I just added a check to verify it.

rpgoldman replied:
It has a different shape because of the change to the predictor variable.  When we reset the `Data` item, we change the shape of the predictor, and so we change the shape of the observed variable.

This is not a change in the function of the code -- I just tested it slightly more thoroughly to verify the prior predictive sampling was working properly (since posterior predictive sampling did not, and does not -- when you change the shape of the predictors, in general, you must get a new model and may need to edit the posterior trace.).

lucianopaz reviewed (changes requested):
I know that you've worked on this for a very long time and I haven't reviewed it in quite a while. My main critique was and still is code complexity. I honestly can't devout much time to pymc3, and I'm focusing the time I have on pymc4 instead, so I don't want to block this PR from being merged. I've left two comments above and once you reply to those, I'll merge this into master, but forward sampling bug fixes from the time of the merge onwards will have to be applied to two independent forward sampling branches

rpgoldman commented:
> I know that you've worked on this for a very long time and I haven't reviewed it in quite a while. My main critique was and still is code complexity. I honestly can't devout much time to pymc3, and I'm focusing the time I have on pymc4 instead, so I don't want to block this PR from being merged. I've left two comments above and once you reply to those, I'll merge this into master, but forward sampling bug fixes from the time of the merge onwards will have to be applied to two independent forward sampling branches

I understand the problem, but I am not sure what to do about it.  There is still the possibility that the two versions of posterior predictive sampling are incomparable, in the sense that they may each have slightly different bug cases.  I don't know what to do about this beyond checking *both* versions each time we introduce a bug fix.

The only thing I can suggest, given everyone's time constraints, is that you try to ensure that forward/predictive sampling not be an afterthought in PyMC4 the way it is in PyMC3.

Particularly of concern: make sure that transformations work properly in both directions, not only in one direction as in PyMC3, and that mixture models can be forward sampled.

End of editorial/sermon! 😉

lucianopaz commented:
Great work @rpgoldman!
